[WIP] transfer learning

b-flo opened the pull request:
This PR adress @hirofumi0810 TODO for transfer learning in `asr/pytorch_backend/asr.py` : 

>better to simplify the E2E model interface by only allowing idim, odim, and args
the pretrained ASR and MT model arguments should be removed here and we should implement additional method to attach these methods

The proposed methods would be used to replace the following in `asr/pytorch_backend/asr.py`: 

```
asr_model, mt_model = None, None

if args.asr_model:
    asr_model, _ = load_trained_model(args.asr_model)
    assert isinstance(asr_model, ASRInterface)

if args.mt_model:
    mt_model, _ = load_trained_model(args.mt_model)
    assert isinstance(mt_model, ASRInterface)

 model_class = dynamic_import(args.model_module)

if asr_model is None and mt_model is None:
    model = model_class(idim, odim, args)
else:
    model = model_class(idim, odim, args, asr_model=asr_model, mt_model=mt_model)
assert isinstance(model, ASRInterface)

if args.asr_model:
    del asr_model

if args.asr_model:
    del asr_model
```

By:

```
if args.enc_init or args.dec_init:
    model = load_trained_modules(idim, odim, args)
else:
    model_class = dynamic_import(args.model_module)
    model = model_class(idim, odim, args)
```

And remove `load_trained_model` method in same file + pre-training code in `nets/pytorch_backend/e2e_asr.py`.

It also extend support by allowing user to specify the modules to transfer from the trained model, for both encoder and decoder. For the decoder part, a pre-trained ESPNET LM is now allowed to support decoder pre-initialization in vanilla transducer.

Before making any changes in  `asr/asr.py` and `nets/e2e_asr.py`, and remove current pre-init arguments (`asr_model` and `mt_model`), I have some questions : 

 - Should the specified trained model(s) architecture overwrite the architecture specified in training config?
 - For now I isolated the methods in `asr/pytorch_backend/asr_init.py`. I don't think I should add them in `asr/pytorch_backend/asr.py` for readibility nor in `asr/asr_utils.py` as it only applies to pytorch. Where should I put them?
 - Does current experiments need additional features such as freezing modules or multi-models support?

sw005320 commented:
Thanks, @b-flo!
Is it possible to make a simple unit test for this initialization with a pre-trained model?

b-flo commented:
Sure! I'll add unit test after review and receiving answer to the first and second question. Right now the code is not covered because the methods are not called in `asr/pytorch_backend/asr.py` and `nets/pytorch_backend/e2e_asr.py`

b-flo commented:
Ultimately I changed the transfer learning implementations in `asr/pytorch_backend/asr.py` and  `nets/pytorch_backend/e2e_asr.py`, and added tests (to extend) in `test/test_asr_init.py`. It may be better for review.

sw005320 commented:
@hirofumi0810, could you review this PR?
This is a very nice extension for you as well.

kan-bayashi commented:
@hirofumi0810 ping

hirofumi0810 commented on test/test_asr_init.py:
```diff
@@ -0,0 +1,166 @@
+    tmppath = tempfile.mktemp()
+    utils.torch_save(tmppath, model)
+
+    if enc_init:
```
if enc_init is not None:

hirofumi0810 commented on test/test_asr_init.py:
```diff
@@ -0,0 +1,166 @@
+
+    if enc_init:
+        enc_init = tmppath
+    if dec_init:
```
if dec_init is not None:

hirofumi0810 commented on espnet/asr/pytorch_backend/asr.py:
```diff
@@ -309,34 +285,14 @@ def train(args):
-    # to attach these models
-    if asr_model is None and mt_model is None:
-        model = model_class(idim, odim, args)
+    if args.enc_init or args.dec_init:
```
if args.enc_init is not None or args.dec_init is not None:

hirofumi0810 commented on espnet/asr/pytorch_backend/asr_init.py:
```diff
@@ -0,0 +1,234 @@
+            modules_model += [(key_m, value_m.shape)]
+
+    len_match = (len(modules_model) == len(partial_modules))
+    module_match = (sorted([x for x in modules_model]) ==
```
Can you specify a key with lambda function?

b-flo replied:
Ha yes, it would be more elegant!

hirofumi0810 commented on espnet/asr/pytorch_backend/asr_init.py:
```diff
@@ -0,0 +1,234 @@
+    module_match = (sorted([x for x in modules_model]) ==
+                    sorted([x for x in partial_modules]))
+
+    return len_match and module_match
```
You seem to skip shape matching. Is it OK? I think I added such parts...

b-flo replied:
I don't skip shape matching, `modules_model` (main) and `partial_modules` (trained)  contains (name, shape) for matched in `modules`. Thus this test will return false if there is a shape mismatch!

b-flo commented:
@hirofumi0810 I addressed your reviews, thanks!

As it seems you'll be using transfer learning, do you think I should add in this PR simple features such as freezing layers/modules? It can be done by modifying `requires_grad` flag from specified parameters and filtering the parameters in optimizer (to avoid updating weights even if grad is 0). 
Also, I was wondering if we should allow the pre-trained model config to overwrite the config for specified modules?

hirofumi0810 commented:
> @hirofumi0810 I addressed your reviews, thanks!
> 
> As it seems you'll be using transfer learning, do you think I should add in this PR simple features such as freezing layers/modules? It can be done by modifying `requires_grad` flag from specified parameters and filtering the parameters in optimizer (to avoid some optimizers to update the weights even if grad is 0).
> Also, I was wondering if we should allow the pre-trained model config to overwrite the config for specified modules?

I think such features are not necessary for this PR. (I'm not using parameter freezing)
With regard to overwriting config files, do you need it for RNN-T?

b-flo commented:
No I don't I was just wondering! It's just that it seems more user-friendly as the user doesn't need to set manually the config file to match pre-trained model config.

sw005320 commented:
I prefer to have a smaller PR. We can always do it later.

sw005320 commented:
I like the current design of splitting encoder/decoder as it would be very useful for RNN-T and ST. But at the same time, many people want to have more straightforward transfer learning for the whole encoder/decoder (e.g.,  we don't change any ASR model architecture and only re-train the parameter values without separating the encoder/decoder for the adaptation or teacher-student learning scenario). Can we easily do it with this approach? I personally feel that we could also have an option to do such things.

hirofumi0810 commented:
@sw005320 We can do that easily because I moved ST from ASR modules in #1139.

sw005320 commented:
OK. Is it ready for merge then? @b-flo and @hirofumi0810 ?

b-flo commented:
If it's okay for @hirofumi0810, the PR can be merged!
Btw if you want to transfer encoder-decoder from the same model, you can specify the same model for `--enc-init` or/and `--dec-init` and keep default `--enc-init-mods` and `dec-init-mods`. But at the same time, depending on the task, even when you want to transfer a complete pre-trained model you want to remove output layer and specific linear layers!

b-flo commented:
As it's related, I'm posting this here for the moment. Early results for transducer with pre-initialization : 

````
db: voxforge-dt_it               CER   |  WER
------------------------------------------------
CTC:                             17.1  |  64.4
Att:                             XX.X  |  XX.X
------------------------------------------------
Trans (no init):                 16.1  |  47.4
Trans (encoder pre-init):        11.7  |  42.2
------------------------------------------------
Trans-att (no init):             14.4  |  46.6
Trans-att (encoder pre-init):    11.7  |  43.3
Trans-att (enc+dec pre-init):    11.1  |  40.9
````

CTC model was trained with the same encoder architecture specified in `egs/voxforge/train_transducer.yaml` (minus dropout).
Outside better overall performances, the observations are in range with the ones I reported in #1129 for my initial experiments : CTC encoder pre-initialization seems more beneficial to vanilla transducer than transducer-attention. However adding decoder pre-initialization with attention-only model further improves transducer-att.
After investigation on why I had degraded performances by using LSTMCell in vanilla transducer, I'll change back and test decoder pre-init with LM to compare.

esalesky commented:
It looks like these updated flags (`enc_init`, `enc-init-mods`, `dec_init`, `dec-init-mods`) for asr/st pretraining didn't make it into mt_train.py, but they're used in the train.yaml files for at least one of the mt recipes (eg must-c). 

Has anyone updated mt_train.py locally (@hirofumi0810 @b-flo ?) 
If not, I can do so and push it.
